csrc/multidevice/collective.h

```diff
@@ -0,0 +1,172 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#pragma once
+#ifdef USE_DISTRIBUTED
+
+#include <multidevice/communicator.h>
+#include <multidevice/multidevice.h>
+
+namespace nvfuser {
+
+/*
+  This struct gathers all the parameters necessary for the
+  construction a collective
+*/
+struct TORCH_CUDA_CU_API CollectiveParams {
+  DeviceIdxType root = -1;
+  std::vector<at::Tensor> src_bufs;
+  std::vector<at::Tensor> dst_bufs;
+  std::vector<DeviceIdxType> team; // should not have duplicate
+};
+
+/*
+The class "Collective" represents a MPI-style collective
+communication operation to be executed on the network. The base class Collective
+should not be used directly but through its derived classes:
+Broadcast, Gather, Scatter, Allgather, and SendRecv. Other collectives will be
+added later.
+
+Later, Collective could be made a derived class of Expr and be thought
+as a kernel IRs resulting of the lowering of a PipelineCommunication.
+
+CollectiveParams contains the arguments for the collective constructors.
+Note that each rank (associated with a device index through
+communicator.deviceId()) will fill CollectiveParams with different arguments,
+depending on the role they play in this collective. For example, the root of a
+Gather collective will have <team_size> destination buffers, whereas non-root
+will have no destination buffers. Also, the ranks not participating in the
```

Comment:
Are these constructors themselves collective operations? In other words, should not-participating ranks call the constructors as well?